subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPacker.java

```diff
@@ -263,12 +265,12 @@ private TarArchiveEntry unpackTree(String treeName, TreeType treeType, File tree
     }
 
     private RegularFileSnapshot unpackFile(TarArchiveInputStream input, TarArchiveEntry entry, File file, String fileName) throws IOException {
-        try (CountingOutputStream output = new CountingOutputStream(new FileOutputStream(file))) {
-            HashCode hash = streamHasher.hashCopy(input, output);
+        try (OutputStream output = new FileOutputStream(file)) {
+            IOUtils.copyLarge(input, output);
```

Comment:
`DefaultStreamHasher` used a pool of buffers to avoid the allocation overhead. I think we should keep that. Perhaps we should separate out the buffer pool as a separate service?

https://github.com/gradle/gradle/blob/9b096a41e2ed9ca69c162acd2682cf3adeec6012/subprojects/hashing/src/main/java/org/gradle/internal/hash/DefaultStreamHasher.java#L65-L76

Reply:
I tried reusing the buffer that's used for storing the tar entry. PTAL